bson-codec-design/bson/bson.go

```diff
@@ -0,0 +1,115 @@
+package bson
+
+import (
+	"io"
+	"reflect"
+	"time"
+
+	"github.com/mongodb/mongo-go-driver/bson-codec-design/bson/decimal"
+	"github.com/mongodb/mongo-go-driver/bson-codec-design/bson/objectid"
+)
+
+type Codec interface {
+	EncodeValue(Registry, ValueWriter, interface{}) error
+	DecodeValue(Registry, ValueReader, interface{}) error
+}
+
+type ArrayReader ValueReader
+
+type DocumentReader interface {
+	ReadElement() (string, ValueReader, error)
+}
+
+type ValueReader interface {
+	ReadBytes([]byte) error
+	ReadSlice() ([]byte, error)
+	Size() int
+	Type() Type
+	Skip() error
+
+	ReadArray() (ArrayReader, error)
+	ReadBinary() (b []byte, btype byte, err error)
+	ReadBoolean() (bool, error)
+	ReadDocument() (DocumentReader, error)
+	ReadCodeWithScope() (code string, dr DocumentReader, err error)
+	ReadDBPointer() (ns string, oid objectid.ObjectID, err error)
+	ReadDateTime() (int64, error)
+	ReadDecimal128() (decimal.Decimal128, error)
+	ReadDouble() (float64, error)
+	ReadInt32() (int32, error)
+	ReadInt64() (int64, error)
+	ReadJavascript() (code string, err error)
+	ReadMaxKey() error
+	ReadMinKey() error
+	ReadNull() error
+	ReadObjectID() (objectid.ObjectID, error)
+	ReadRegex() (pattern, options string, err error)
+	ReadString() (string, error)
+	ReadSymbol() (symbol string, err error)
+	ReadTime() (time.Time, error)
```

Comment:
It's so that you can make a timestamp with a time.Time instance instead of 2 uint32s.

Reply:
I see that as the job of the TImeCodec, which can be implemented in terms of ReadDateTime, and perhaps configured with a default Location

Reply:
I think this makes sense for parity with the constructors and for convenience of users who are implementing a codec, why should I need to lookup another codec to get a time.Time object? That's like requiring another codec for decimal128 or objectID.

Reply:
I would expect that we would have (very simple) codecs for those too, as it makes implementation of higher level codecs (like for structs) easy to write: just look up a codec in the registry for whatever type you encounter.

Happy to defer this to implementation though.

Reply:
The problem is then the ValueReader and ValueWriter don't work with the ExtJSON and `*bson.Document* variants, because there won't be a `[]byte` or other representation that we can really give to another codec. It's a bootstrapping issue, and `time.Time` is the only one we can bootstrap from timestamp.

Reply:
I don't understand how the other variants are relevant here.  We'll have to discuss.

Reply:
Agreed to remove it.